Add linkText to AlphabetHints

danielcovill opened the pull request:
This should make it easier to discern scroll/frame markers from links. It's a feature that has been only available with Filter hints enabled. I may be missing something, but I see no reason why it shouldn't be visible when using the default Alphabet hints as well. 

https://github.com/philc/vimium/issues/4675

danielcovill committed:
Add linkText to AlphabetHints so it's easier to discern scroll/frame markers

philc commented:
I'll look at this soon.

On Wed, Jun 25, 2025 at 10:43 AM, Jan-Henrik Damaschke < ***@***.*** > wrote:

> 
> *itpropro* left a comment (philc/ vimium#4689) (
> https://github.com/philc/vimium/pull/4689#issuecomment-3005637545 )
> 
> 
> @ philc ( https://github.com/philc ) any feedback?
> 
> 
> 
> —
> Reply to this email directly, view it on GitHub (
> https://github.com/philc/vimium/pull/4689#issuecomment-3005637545 ) , or unsubscribe
> (
> https://github.com/notifications/unsubscribe-auth/AAACDFUUMYPZSJYIY436PJT3FLNUZAVCNFSM6AAAAAB36B46Z6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMBVGYZTONJUGU
> ).
> You are receiving this because you were mentioned. Message ID: <philc/vimium/pull/4689/c3005637545
> @ github. com>
> 
> 
>

philc commented:
UX comments:

Adding "scroll" and "frame" to the hint text can add a lot of clutter to the page and obscure nearby hints. The UX rationale for only having it in filtered hints has been because the clutter in that mode is less of a problem, because you can very quickly type a few characters that you see in the anchor, or in the hint itself. With alphabet hints, you can't quickly narrow out the clutter like this.

Second, showing "scroll" and "frame" in filtered hints isn't my favorite because it's not self-explanatory. But at least that mode has the convention of showing the name of some links in the hints (e.g. for form controls). In alphabet hints, having the word "scroll" appear next to a hint is even less self-explanatory, and more unusual. I think a change in color might work better, or some other UX. Any thoughts?

Feedback about this PR:

1. Which page are you primarily testing this on?
2. If this PR, there should only be link text shown for frames and scrollables. I noticed the values of the form controls on the Vimium options page are also showing link text. See screenshot.
<img width="575" alt="Screenshot 2025-06-28 at 6 36 07 PM" src="https://github.com/user-attachments/assets/bab7a02d-7183-4de3-8f77-fd50e9ffb96e" />

I found an adjacent bug which is unrelated to this PR. I tested this change on test_harnesses/iframe.html in the Vimium repo. After hitting space to toggle the hints, typing S navigates to the first link starting with an S, but instead it should wait for further characters to be typed.

philc commented on content_scripts/link_hints.js:
```diff
@@ -792,18 +792,21 @@ class AlphabetHints {
 
+  renderMarker(marker) {
+    let linkText = marker.linkText;
+    console.log(linkText);
```
Log statement left in.

philc commented on content_scripts/link_hints.js:
```diff
@@ -792,18 +792,21 @@ class AlphabetHints {
+
   fillInMarkers(hintMarkers) {
     const hintStrings = this.hintStrings(hintMarkers.length);
-    if (hintMarkers.length != hintStrings.length) {
```
Is this safety check no longer needed?

Commit:
Merge branch 'master' into highlight-scroll-hints

Commit:
Enhance link hinting with scroll marker and overlap handling

- Added 'scroll-hint-marker' class for frames and scrollable areas in link hints.
- Implemented overlap detection and offset adjustments for overlapping hint markers.
- Updated CSS for scroll hint markers to improve visibility.

danielcovill commented:
- Thanks for the callouts on the review, both the log statement and safety check were sloppy mistakes. I've addressed those.
- I am now testing using test_harnesses/iframe.html and the Vimium options page.
- I revisited and changed the approach based on your comments. When a link is "Scroll." or "Frame." I'm changing the color of the hint-marker. 
- Additionally, I found that the hint markers for scroll/frame frequently end up stacked behind other markers so I added a bit of code to check for that scenario and add a slight offset (2px down and right) to any markers that are in the exact same spot. I think it provides a nice indication that there are markers behind one another. 

Happy to remove that last bit if it exceeds the scope of this PR.

icalvo commented:
My vote for accepting this PR, telling apart scroll/frame markers is important in pages with lots of scrollable elements. Also, `gf` doesn't work on scrollable elements so `f` is the only way to activate and scroll them.

danielcovill commented:
Hi @philc - I'd love a review if possible, please. Thanks
